File: archive/parallel-learning/src/goals/goal.py
"""目标数据模型 — 学习目标的状态和数据结构"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional


class GoalStatus(Enum):
    """目标状态"""
    PENDING = 'pending'
    IN_PROGRESS = 'in_progress'
    COMPLETED = 'completed'
    BLOCKED = 'blocked'
    ABANDONED = 'abandoned'


@dataclass
class Goal:
    """学习目标

    一个可分解、可追踪的学习目标。
    """
    id: str
    description: str
    status: GoalStatus = GoalStatus.PENDING
    sub_goals: List[str] = field(default_factory=list)
    parent_goal: Optional[str] = None
    required_knowledge: List[str] = field(default_factory=list)
    priority: float = 0.5
    progress: float = 0.0
    created_step: int = 0

    def to_dict(self) -> dict:
        """序列化为字典"""
        return {
            'id': self.id,
            'description': self.description,
            'status': self.status.value,
            'sub_goals': self.sub_goals,
            'parent_goal': self.parent_goal,
            'required_knowledge': self.required_knowledge,
            'priority': self.priority,
            'progress': self.progress,
            'created_step': self.created_step,
        }

    @classmethod
    def from_dict(cls, data: dict) -> 'Goal':
        """从字典反序列化"""
        status_val = data.get('status', 'pending')
        if isinstance(status_val, str):
            status_val = GoalStatus(status_val)
        return cls(
            id=data['id'],
            description=data['description'],
            status=status_val,
            sub_goals=data.get('sub_goals', []),
            parent_goal=data.get('parent_goal'),
            required_knowledge=data.get('required_knowledge', []),
            priority=data.get('priority', 0.5),
            progress=data.get('progress', 0.0),
            created_step=data.get('created_step', 0),
        )

File: ai-learning-cpp/src/goals/goal.cpp
/**
 * @file goal.cpp
 * @brief Goal 值对象序列化 + 枚举转换
 */

#include "ai_learning/goals/goal.hpp"

#include <algorithm>
#include <sstream>

namespace ai_learning::goals {

// ── 枚举 → 字符串 ──────────────────────────────────────────────

auto goal_status_to_string(GoalStatus s) -> std::string {
    switch (s) {
        case GoalStatus::Pending:    return "pending";
        case GoalStatus::InProgress: return "in_progress";
        case GoalStatus::Completed:  return "completed";
        case GoalStatus::Blocked:    return "blocked";
        case GoalStatus::Abandoned:  return "abandoned";
    }
    return "pending";
}

auto string_to_goal_status(const std::string& s) -> GoalStatus {
    if (s == "in_progress") return GoalStatus::InProgress;
    if (s == "completed")   return GoalStatus::Completed;
    if (s == "blocked")     return GoalStatus::Blocked;
    if (s == "abandoned")   return GoalStatus::Abandoned;
    return GoalStatus::Pending;
}

auto strategy_to_string(LearningStrategy s) -> std::string {
    switch (s) {
        case LearningStrategy::Decompose:  return "decompose";
        case LearningStrategy::Analogize:  return "analogize";
        case LearningStrategy::Practice:   return "practice";
        case LearningStrategy::Explore:    return "explore";
    }
    return "explore";
}

auto action_to_string(StepAction a) -> std::string {
    switch (a) {
        case StepAction::Learn:    return "learn";
        case StepAction::Execute:  return "execute";
    }
    return "learn";
}

// ── 辅助：逗号连接 ──────────────────────────────────────────────

static auto join_comma(const std::vector<std::string>& v) -> std::string {
    std::string r;
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) r += ",";
        r += v[i];
    }
    return r;
}

static auto split_comma(const std::string& s) -> std::vector<std::string> {
    std::vector<std::string> result;
    std::istringstream iss(s);
    std::string token;
    while (std::getline(iss, token, ',')) {
        // 去除首尾空格
        auto start = token.find_first_not_of(' ');
        auto end   = token.find_last_not_of(' ');
        if (start != std::string::npos && end != std::string::npos) {
            result.push_back(token.substr(start, end - start + 1));
        }
    }
    return result;
}

// ── Goal::to_map ────────────────────────────────────────────────

auto Goal::to_map() const -> std::map<std::string, std::string> {
    return {
        {"id",                 id},
        {"description",        description},
        {"status",             goal_status_to_string(status)},
        {"sub_goals",          join_comma(sub_goals)},
        {"parent_goal",        parent_goal},
        {"required_knowledge", join_comma(required_knowledge)},
        {"priority",           std::to_string(priority)},
        {"progress",           std::to_string(progress)},
        {"created_step",       std::to_string(created_step)},
    };
}

// ── Goal::from_map ──────────────────────────────────────────────

auto Goal::from_map(const std::map<std::string, std::string>& m) -> Goal {
    Goal g;
    g.id          = m.count("id")          ? m.at("id")          : "";
    g.description = m.count("description") ? m.at("description") : "";
    g.status      = m.count("status")      ? string_to_goal_status(m.at("status"))
                                           : GoalStatus::Pending;
    g.sub_goals          = m.count("sub_goals")          ? split_comma(m.at("sub_goals"))          : std::vector<std::string>{};
    g.parent_goal        = m.count("parent_goal")        ? m.at("parent_goal")                     : "";
    g.required_knowledge = m.count("required_knowledge") ? split_comma(m.at("required_knowledge")) : std::vector<std::string>{};
    g.priority     = m.count("priority")     ? std::stod(m.at("priority"))     : 0.5;
    g.progress     = m.count("progress")     ? std::stod(m.at("progress"))     : 0.0;
    g.created_step = m.count("created_step") ? std::stoi(m.at("created_step")) : 0;
    return g;
}

}  // namespace ai_learning::goals
